packages/permissions-kernel-snap/src/utils/validate.ts

```diff
@@ -55,3 +96,79 @@ export const parsePermissionsResponseParam = (
 
   return validatePermissionsResponse.data;
 };
+
+/**
+ * Zod schema for validating JSON-RPC request structure
+ */
+export const zJsonRpcRequest = z.object({
+  /**
+   * The JSON-RPC version, must be "2.0"
+   */
+  jsonrpc: z.literal('2.0'),
+
+  /**
+   * The method name - must be a valid RPC method
+   */
+  method: z.nativeEnum(RpcMethod),
+
+  /**
+   * The parameters for the method - must be valid JsonRpcParams
+   */
+  params: z
+    .union([
+      z.string(),
+      z.number(),
+      z.boolean(),
+      z.null(),
+      z.array(z.unknown()).refine((arr) => arr.every(isSafeObject), {
+        message: 'Invalid key in array: potential prototype pollution attempt',
+      }),
+      z.record(z.unknown()).refine((obj) => isSafeObject(obj), {
+        message: 'Invalid key: potential prototype pollution attempt',
+      }),
+    ])
+    .optional(),
+
+  /**
+   * The request ID - must be a string or number
+   */
+  id: z.union([z.string(), z.number()]).optional(),
+});
+
+/**
+ * Type for a validated JSON-RPC request
+ */
+export type ValidatedJsonRpcRequest = z.infer<typeof zJsonRpcRequest>;
+
+/**
+ * Validates that the request object is a proper JSON-RPC request
+ * and that the method is supported by this snap.
+ *
+ * @param request - The request object to validate.
+ * @returns The validated request object.
+ * @throws InvalidParamsError if validation fails.
+ */
+export function validateJsonRpcRequest(
+  request: unknown,
+): ValidatedJsonRpcRequest {
+  // Validate the JSON-RPC structure using Zod
+  const validationResult = zJsonRpcRequest.safeParse(request);
+
+  if (!validationResult.success) {
+    const errorMessages = validationResult.error.errors.map((error) => {
+      const path = error.path.length > 0 ? error.path.join('.') : 'root';
+      return `${path}: ${error.message}`;
+    });
```

Comment:
I think we already have an `extractZodError` helper in `packages/shared/src/utils/error.ts` for handling Zod error messages.Perhaps we could reuse it here?